qstarz_bl_1000.cc

```diff
@@ -0,0 +1,315 @@
+  quint32               unused2;                  // 60-64
+
+  stream >> fixStatus;
+  stream >> rcr;
+  stream >> milliseconds;
+
+  stream.setFloatingPointPrecision(QDataStream::DoublePrecision);
+  stream >> dLatitude;
+  stream >> dLongitude;
+
+  stream >> time;
+
+  stream.setFloatingPointPrecision(QDataStream::SinglePrecision);
+  stream >> speed;
+  stream >> altitude;
+  stream >> heading;
+
+  stream >> gx;
+  stream >> gy;
+  stream >> gz;
+  stream >> maxSNR;
+
+  stream.setFloatingPointPrecision(QDataStream::SinglePrecision);
+  stream >> hdop;
+  stream >> vdop;
+
+  stream >> satelliteCountView;
+  stream >> satelliteCountUsed;
+  stream >> fixQuality;
+  stream >> batteryPercent;
+  stream >> unused1;
+  stream >> unused2;
+
+  if (stream.status() != QDataStream::Ok) {
+    fatal(MYNAME ":  error on '%s'. Perhaps this isn't a Qstarz BL-1000 file\n", qPrintable(read_fname));
+  }
+
+  BL1000_POINT_TYPE type;
+
+  switch (rcr) {
+  case 'B':
+    type = BL1000_POINT_TYPE_WAY_POINT;
+    break;
+  case 'T':
+    type = BL1000_POINT_TYPE_TRACK_POINT_BY_TIME;
+    break;
+  case 'D':
+    type = BL1000_POINT_TYPE_TRACK_POINT_BY_DISTANCE;
+    break;
+  case 'S':
+    type = BL1000_POINT_TYPE_TRACK_POINT_BY_SPEED;
+    break;
+  case 'W': // Power On
+  case 'X': // Power Off
+  case 'Y': // USB plug in
+  case 'Z': // USB removed
+  case 'C': // Cali G-sensor
+  case 'A': // Enter Smart Mode(Power Saving)
+  case 'R': // Reserved
+  case 'N': // Reserved
+  case 'M': // Reserved
+  case 'G': // Reserved
+  case 'H': // Reserved
+  case 'I': // Reserved
+  case 'J': // Reserved
+    type = BL1000_POINT_TYPE_UNKNOWN;
+    break;
+  default:
+    type = BL1000_POINT_TYPE_UNKNOWN;
+
+    fatal(MYNAME ": File format error on '%s'. Unexpected value for RCR (record reason): %c\n", qPrintable(read_fname), (char)rcr);
+
+    break;
+  }
+
+  fix_type fix;
+
+  switch (fixQuality) {
+  case 1:
+    switch (fixStatus) {
+    case 1:
+      fix = fix_none;
+      break;
+    case 2:
+      fix = fix_2d;
+      break;
+    case 3:
+      fix = fix_3d;
+      break;
+    default:
+      fix = fix_unknown;
+      break;
+    }
+    break;
+  case 2:
+    fix = fix_dgps;
+    break;
+  case 3:
+    fix = fix_pps;
+    break;
+  default:
+    fix = fix_unknown;
+
+    if (type != BL1000_POINT_TYPE_UNKNOWN) {
+      fatal(MYNAME ": File format error on '%s'. Unexpected value for fix quality: %d\n", qPrintable(read_fname), fixQuality);
+    }
+
+    break;
+  }
+
+  auto* waypoint = new Waypoint();
+
+  int tmp_lat = dLatitude / 100;
+  int tmp_lon = dLongitude / 100;
+
+  waypoint->latitude = tmp_lat + (dLatitude - tmp_lat * 100) / 60.0;
+  waypoint->longitude = tmp_lon + (dLongitude - tmp_lon * 100) / 60.0;
+
+  // printf(MYNAME ": waypoint->latitude: %f\n", waypoint->latitude);
+  // printf(MYNAME ": waypoint->longitude: %f\n", waypoint->longitude);
+
+  if ((waypoint->latitude < -90) || (waypoint->latitude > 90)) {
```

Comment:
Good to know. I prefer to leave such sanity checks here. Many file formats use a .BIN extension. When we read values that are out of range, I conclude that we must be looking at "random data" instead of an actual QStarz file